crates/solidity/outputs/cargo/tests/src/bindings_output/runner.rs

```diff
@@ -0,0 +1,234 @@
+use std::fmt;
+use std::ops::Range;
+use std::path::Path;
+
+use anyhow::Result;
+use ariadne::{Color, Config, Label, Report, ReportBuilder, ReportKind, Source};
+use infra_utils::cargo::CargoWorkspace;
+use infra_utils::codegen::CodegenFileSystem;
+use infra_utils::github::GitHub;
+use infra_utils::paths::PathExtensions;
+use metaslang_bindings::builder;
+use metaslang_graph_builder::ast::File;
+use metaslang_graph_builder::graph::{Graph, Value};
+use metaslang_graph_builder::{ExecutionConfig, NoCancellation, Variables};
+use semver::Version;
+use slang_solidity::bindings::{self, Handle};
+use slang_solidity::cst::KindTypes;
+use slang_solidity::language::Language;
+use slang_solidity::parse_output::ParseOutput;
+
+use super::generated::VERSION_BREAKS;
+
+pub fn run(group_name: &str, test_name: &str) -> Result<()> {
+    let test_dir = CargoWorkspace::locate_source_crate("solidity_testing_snapshots")?
+        .join("bindings_output")
+        .join(group_name)
+        .join(test_name);
+
+    let mut fs = CodegenFileSystem::new(&test_dir)?;
+
+    let input_path = test_dir.join("input.sol");
+
+    let source = input_path.read_to_string()?;
+
+    let mut last_graph_output = None;
+    let mut last_bindings_output = None;
+
+    for version in &VERSION_BREAKS {
+        let language = Language::new(version.clone())?;
+
+        let parse_output = language.parse(Language::ROOT_KIND, &source);
+
+        if !GitHub::is_running_in_ci() {
+            // Don't run this in CI, since the graph outputs are not committed
+            // to the repository and hence we cannot verify their contents,
+            // which is what `fs.write_file` does in CI.
+            let graph_output = output_graph(version, &parse_output)?;
+            match last_graph_output {
+                Some(ref last) if last == &graph_output => (),
+                _ => {
+                    let snapshot_path = test_dir.join("generated").join(format!("{version}.mmd"));
+
+                    fs.write_file(snapshot_path, &graph_output)?;
+                    last_graph_output = Some(graph_output);
+                }
+            };
+        }
+
+        let bindings_output = output_bindings(version, &parse_output, &source, &input_path)?;
+        match last_bindings_output {
+            Some(ref last) if last == &bindings_output => (),
+            _ => {
+                let snapshot_path = test_dir.join("generated").join(format!("{version}.txt"));
+
+                fs.write_file(snapshot_path, &bindings_output)?;
+                last_bindings_output = Some(bindings_output);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+const ROOT_NODE_VAR: &str = "ROOT_NODE";
+const VERSION_VAR: &str = "VERSION";
+
+const VARIABLE_DEBUG_ATTR: &str = "__variable";
+const LOCATION_DEBUG_ATTR: &str = "__location";
+const MATCH_DEBUG_ATTR: &str = "__match";
+
+fn output_graph(version: &Version, parse_output: &ParseOutput) -> Result<String> {
```

Comment:
Yes, this is a problem indeed. Also, whenever we change the conditions for building the graph, we need to make changes in two places. This graph is only for debugging purposes though, so it's not as critical except for developing the rules.

The issue with using the one from `Bindings` is that there's no easy way to extract it. The graph is built from the source and then fed into the stack graph, but that's as far as it lives.

Reply:
I don't see an easy way to fix this. In some sense, this is related to your [other comment](https://github.com/NomicFoundation/slang/pull/1036#discussion_r1676806805) about exposing the binding rules. In both cases, we need to access some of the internals of the bindings to test them or to improve the DX. And the other thing is that those internals are in a different crate as the one with the tests. What do you think about using a (hidden?) feature flag for this, so that only the test crate would enable it?

Reply:
I changed this and exposed an additional method in `Bindings` to add a file to the bindings but also return the intermediate graph, so we can print it here. This new method is hidden behind the `__private_testing_utils` feature flag.